### DISCUSSION TITLE
How to preserve dataset order when using DDP?

### DISCUSSION THREAD

**Discussion by malfonsoarquimea** [Lightning-AI/pytorch-lightning#15164] [DDP / multi-GPU / multi-node] [answered]
I need to be able to preserve the order in which the data is fed to the model when training in multiple GPUS.
According to https://github.com/Lightning-AI/lightning/discussions/13342 each GPU gets a consecutive fraction of the dataset, so if I have 2GPUs, the first one will get the first half of the dataset and the other one will get the second half of the dataset.
I need to preserve the order and don't know how to overwrite the dataset-splitting logic. Any advice?

**Comment by rohitgr7** [comment, marked as answer]
for that you'd need to write a custom [DistributedSampler](https://pytorch.org/docs/stable/_modules/torch/utils/data/distributed.html#DistributedSampler) and pass it to the dataloader and set `Trainer(replace_sampler_ddp=False)`

**Reply by malfonsoarquimea** [reply]
thanks, tested it and working perfectly!

**Reply by rohitgr7** [reply]
is it possible for you to share your `DistributedSampler` implementation here so that others can benefit from it?

**Reply by malfonsoarquimea** [reply]
I just used the default DistributedSampler
```
 def train_dataloader(self):
        """returns a dataloader for training according to hparams

        Returns:
            DataLoader: DataLoader ready to deliver samples for training
        """
        # define a distributed sampler in case we are using multiple GPUs
        if self.hparams.num_gpus>1:
            sampler = torch.utils.data.distributed.DistributedSampler(
                self.train_dataset, shuffle=False)
        # only use the sampler if using multiple GPUs
        return DataLoader(
            self.train_dataset,
            shuffle=False,
            num_workers=self.hparams.num_workers,
            batch_size=self.hparams.batch_size,
            pin_memory=False,
            sampler=sampler if self.hparams.num_gpus > 1 else None)
```

**Reply by Hannibal046** [reply]
Hi, @rohitgr7 , In this case, do I have to take care of `single-gpu` and `multi-gpu` training myself ?
